### PR TITLE
Add a file-table-tform tool: manipulating tabular string data files.

### PR DESCRIPTION
**Pull request by kbobrovs**
Simple tool for manipulating tables on files - inserting/replacing columns, adding rows, etc. To be
used to model multiple inputs/outputs in clang driver action graph node in SYCL offload compilations.

The first commit is a prerequisite and is a separate PR #1358

part of #1356

Signed-off-by: Konstantin S Bobrovsky <konstantin.s.bobrovsky@intel.com>

### CODE REVIEW

**Comment by kbobrovs** [comment]
@Fznamznon, @mdtoguchi, @sndmitriev, @AGindinson - please review/approve. All review comments have been addressed.

File: llvm/tools/file-table-tform/file-table-tform.cpp [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+            .Case(OPT_RENAME, [&](TformCmd *Cmd) { return Error::success(); })
+            .Case(OPT_EXTRACT, [&](TformCmd *Cmd) { return Error::success(); });
+    return F(this);
+  }
+
+  Error parseArg(StringRef Arg) {
+    Func F =
+        StringSwitch<Func>(Kind)
+            // need -> Error as lambda return type can't be deduced
+            // automatically
```
Comment (Fznamznon) [inline llvm/tools/file-table-tform/file-table-tform.cpp]: It's unclear what this comment means

File: llvm/tools/file-table-tform/file-table-tform.cpp [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+  }
+
+  Error execute(util::SimpleTable &Table) {
+    Func F =
+        StringSwitch<Func>(Kind)
+            .Case(OPT_REPLACE,
+                  [&](TformCmd *Cmd) -> Error {
+                    // argument is <column name>
+                    assert(Cmd->Args.size() == 2 && Cmd->Inputs.size() == 1);
+                    auto Table1 = util::SimpleTable::read(Cmd->Inputs[0]);
```
Comment (Fznamznon) [inline llvm/tools/file-table-tform/file-table-tform.cpp]: Please do not use `auto` here

File: llvm/tools/file-table-tform/file-table-tform.cpp [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+  auto Lists = {std::addressof(TformReplace), std::addressof(TformRename),
+                std::addressof(TformExtract)};
+
+  for (const auto *L : Lists) {
+    for (auto It = L->begin(); It != L->end(); It++) {
+      auto Cmd = TformCmd::create(L->ArgStr, *It);
+
+      if (!Cmd)
+        CHECK_AND_EXIT(Cmd.takeError());
+      const auto Pos = L->getPosition(It - L->begin());
```
Comment (Fznamznon) [inline llvm/tools/file-table-tform/file-table-tform.cpp]: Please do not use `auto` here

File: llvm/tools/file-table-tform/file-table-tform.cpp [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+  for (auto &P : Cmds) {
+    auto &Cmd = P.second;
+    // this will advance cur iterator as far as needed
+    auto E = Cmd->consumeInput(CurInput, EndInput);
+    CHECK_AND_EXIT(E);
+  }
+  // commands are constructed, command line is correct - read input and execute
+  // transformations on it
+
+  auto Table = util::SimpleTable::read(InputFile);
```
Comment (Fznamznon) [inline llvm/tools/file-table-tform/file-table-tform.cpp]: Please do not use auto here

File: llvm/tools/file-table-tform/file-table-tform.cpp [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+  const auto EndInput = Inputs.end();
+  // first input is the "current" - it will undergo the transformation sequence
+  if (CurInput == EndInput)
+    CHECK_AND_EXIT(makeUserError("no inputs"));
+  std::string &InputFile = *CurInput++;
+
+  for (auto &P : Cmds) {
+    auto &Cmd = P.second;
+    // this will advance cur iterator as far as needed
+    auto E = Cmd->consumeInput(CurInput, EndInput);
```
Comment (Fznamznon) [inline llvm/tools/file-table-tform/file-table-tform.cpp]: Please do not use auto here

File: llvm/tools/file-table-tform/file-table-tform.cpp [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+  // is valid
+  auto CurInput = Inputs.begin();
+  const auto EndInput = Inputs.end();
+  // first input is the "current" - it will undergo the transformation sequence
+  if (CurInput == EndInput)
+    CHECK_AND_EXIT(makeUserError("no inputs"));
+  std::string &InputFile = *CurInput++;
+
+  for (auto &P : Cmds) {
+    auto &Cmd = P.second;
```
Comment (Fznamznon) [inline llvm/tools/file-table-tform/file-table-tform.cpp]: Please do not use auto here

File: llvm/tools/file-table-tform/file-table-tform.cpp [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+  }
+  // commands are constructed, command line is correct - read input and execute
+  // transformations on it
+
+  auto Table = util::SimpleTable::read(InputFile);
+  if (!Table)
+    CHECK_AND_EXIT(Table.takeError());
+
+  for (auto &P : Cmds) {
+    auto &Cmd = P.second;
```
Comment (Fznamznon) [inline llvm/tools/file-table-tform/file-table-tform.cpp]: Please do not use auto here

File: llvm/tools/file-table-tform/file-table-tform.cpp [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// The tool for now supports only linear transformation sequences like
+//      File1       File3
+//         \          \
+// Input - op1 - op2 - op3 - Output
+//         /
+//      File2
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply llvm/tools/file-table-tform/file-table-tform.cpp]: Not clear what specifically is very hard to understand - graph of files and operations. Will try to add more comments

File: llvm/tools/file-table-tform/file-table-tform.cpp [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/Twine.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Errc.h"
+#include "llvm/Support/InitLLVM.h"
+#include "llvm/Support/SimpleTable.h"
+#include "llvm/Support/WithColor.h"
+
+#include <algorithm>
+#include <fstream>
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply llvm/tools/file-table-tform/file-table-tform.cpp]: ok

File: llvm/tools/file-table-tform/file-table-tform.cpp [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+  using UPtrTy = std::unique_ptr<TformCmd>;
+  StringRef Kind;
+  SmallVector<StringRef, 2> Args;
+  SmallVector<StringRef, 2> Inputs;
+
+  TformCmd() = default;
+  TformCmd(StringRef Kind) : Kind(Kind) {}
+
+  static Expected<UPtrTy> create(StringRef Kind, StringRef RawArg = "") {
+    UPtrTy Res(new TformCmd(Kind));
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply llvm/tools/file-table-tform/file-table-tform.cpp]: ok

File: llvm/tools/file-table-tform/file-table-tform.cpp (hunk 1 above)
Comment (kbobrovs) [reply llvm/tools/file-table-tform/file-table-tform.cpp]: ok, will expand

File: llvm/tools/file-table-tform/file-table-tform.cpp (hunk 2 above)
Comment (kbobrovs) [reply llvm/tools/file-table-tform/file-table-tform.cpp]: ok

File: llvm/tools/file-table-tform/file-table-tform.cpp [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+  std::map<int, TformCmd::UPtrTy> Cmds;
+
+  // Partially construct commands (w/o input information). Can't fully construct
+  // yet, as an order across all command line options-commands needs to be
+  // established first to properly map inputs to commands.
+
+  // Construct commands represented by an option with string argument,
+  // auto-sorting them by their position in the command line
+  // Lists[1]&TformRename);
+  // Lists.push_back(&TformExtract);
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply llvm/tools/file-table-tform/file-table-tform.cpp]: Will clean up

File: llvm/tools/file-table-tform/file-table-tform.cpp (hunk 3 above)
Comment (kbobrovs) [reply llvm/tools/file-table-tform/file-table-tform.cpp]: ok

File: llvm/tools/file-table-tform/file-table-tform.cpp (hunk 6 above)
Comment (kbobrovs) [reply llvm/tools/file-table-tform/file-table-tform.cpp]: ok

File: llvm/tools/file-table-tform/file-table-tform.cpp (hunk 5 above)
Comment (kbobrovs) [reply llvm/tools/file-table-tform/file-table-tform.cpp]: ok

File: llvm/tools/file-table-tform/file-table-tform.cpp (hunk 4 above)
Comment (kbobrovs) [reply llvm/tools/file-table-tform/file-table-tform.cpp]: ok

File: llvm/tools/file-table-tform/file-table-tform.cpp (hunk 7 above)
Comment (kbobrovs) [reply llvm/tools/file-table-tform/file-table-tform.cpp]: ok

File: llvm/tools/file-table-tform/file-table-tform.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (kbobrovs) [reply llvm/tools/file-table-tform/file-table-tform.cpp]: OK, I see that the comment about what the tool does is entirely missing. Will add.

**Comment by kbobrovs** [comment]
> kbobrovs force-pushed the kbobrovs:spec-const-table-tform branch from 8d7393d to 7e07365

addressed review comments from @Fznamznon

File: llvm/tools/file-table-tform/file-table-tform.cpp [hunk 12]
```diff
@@ -0,0 +1,322 @@
+//===- file-table-tform.cpp - transform files with tables of strings ------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// This tool transforms a series of input file tables into single output file
+// table according to operations passed on the command line. Operations'
```
Comment (Fznamznon) [inline llvm/tools/file-table-tform/file-table-tform.cpp]:
```suggestion
// table according to operations passed on the command line. Operation's
```

File: llvm/tools/file-table-tform/file-table-tform.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (Fznamznon) [reply llvm/tools/file-table-tform/file-table-tform.cpp]: Feels better now, thanks.

File: llvm/tools/file-table-tform/file-table-tform.cpp (hunk 12 above)
Comment (kbobrovs) [reply llvm/tools/file-table-tform/file-table-tform.cpp]: Operations' = arguments of operations, so English is correct here, as far as I can tell

File: llvm/tools/file-table-tform/file-table-tform.cpp [hunk 13]
```diff
@@ -0,0 +1,322 @@
...
+// arguments are input files, and some operations like 'rename' take single
+// input, others like 'replace' take two. Operations are executed in
+// command-line order and consume needed amount of inputs lef-to-right in the
```
[the remark this answers is not in the document]
Comment (kbobrovs) [reply llvm/tools/file-table-tform/file-table-tform.cpp]: ok

**Comment by bader** [comment]
@kbobrovs, could you resolve the conflict, please?

**Comment by kbobrovs** [comment]
> @kbobrovs, could you resolve the conflict, please?

sure, 1 sec

**Comment by kbobrovs** [comment]
>  kbobrovs force-pushed the kbobrovs:spec-const-table-tform branch from 7e0dc74 to 94d3cd2 19 seconds ago

rebased onto intel/llvm/sycl

File: llvm/tools/file-table-tform/LLVMBuild.txt [hunk 14]
```diff
@@ -0,0 +1,21 @@
+;===- ./tools/llvm-table/LLVMBuild.txt -------------------------*- Conf -*--===;
```
Comment (bader) [inline llvm/tools/file-table-tform/LLVMBuild.txt]:
```suggestion
;===- ./tools/file-table-tform/LLVMBuild.txt -------------------*- Conf -*--===;
```

**Pull request merged**